en/docs/extended_topics/user_op.md

```diff
@@ -1,414 +1,816 @@
+    const user_op::Tensor *in_tensor = ctx->Tensor4ArgNameAndIndex("in", 0);
+    user_op::Tensor *out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    MyRelu<T>(ctx->device_ctx(),
+           in_tensor->shape().elem_cnt(),
+           in_tensor->dptr<T>(), 
+           out_tensor->mut_dptr<T>());
+  }
+  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+};
+
+#define REGISTER_RELU_KERNEL(device, dtype)          \
+  REGISTER_USER_KERNEL("myrelu")                     \
+      .SetCreateFn<ReluKernel<device, dtype>>()      \
+      .SetIsMatchedHob(                              \
+          (user_op::HobDeviceType() == device) &     \
+          (user_op::HobDataType("out", 0)            \
+            == GetDataType<dtype>::value));
+
+REGISTER_RELU_KERNEL(DeviceType::kCPU, float)
+REGISTER_RELU_KERNEL(DeviceType::kCPU, double)
+} // namespace
+
+} // namespace oneflow
 ```
 
-分析下上述代码：
+To implement the kernel in OneFlow, you must define a class which inherits from `oneflow::user_op::OpKernel` and rewrite the virtual functions of it.
 
-首先是include了`oneflow/core/framework/framework.h`。`framework.h`中包括了我们创建一个Op需要的所有头文件，所以我们只需要include这一个头文件就行。
+In the above code, we rewrite `Compute` and `AlwaysComputeWhenAllOutputsEmpty` and their respective meanings are:
 
-其次，为了简化类型名，我们在`namespace oneflow`下书写注册Op定义的相关代码。
+* `Compute` must be rewritten to implement the specific operating logic.
 
-最后，我们来看下注册的代码。`Relu`  op只有一个输入`in`，也只有一个输出`out`，其推导out tensor的形状的`ShapeInferFn`是一个lambda函数，参数为`InferContext`，返回值为`Maybe<void>`类型对象。
+* `AlwaysComputeWhenAllOutputsEmpty` must be rewritten to return `false` for the most of op. For the some op that needs to maintain state internally. Therefore needs to be computed by calling the kernel even if the output is null and return `true`.
 
-具体到`Relu` op对应的`ShapeInferFn`，其输出的形状与输入相同，即如上述代码所示。
+After implementing the kernel class, you need to call `REGISTER_USER_KERNEL` to register it. The string parameter that `REGISTER_USER_KERNEL("myrelu")` accepts is `op_type_name` which is used to complete registration and querying. You also need to use `op_type_name` when wrapping op at the Python layer.
 
+`REGISTER_USER_KERNEL("myrelu") ` returns an `OpKernelRegistry` object. The methods that need to be called to set the registration information are  mention in the code above. 
 
+* `SetCreateFn<T>()`: The method of this template's parameter `T`  is our implementation of the kernel class which OneFlow will use it to create the kernel object.
 
-## 实现Kernel
-在写完Op的定义之后，我们需要为其提供一个或多个Kernel作为实现。此处我们实现一个支持在gpu上执行float数据类型的Kernel，存放在`relu_gpu.cu`文件中。
+* `SetIsMatchedHob`：Because an op may have more than one kernel. You need to call `SetIsMatchedHob` to select a different kernel for the calculation according to the physical device and data format. This method accepts an equation and when the equation is `true`, OneFlow will call the kernel to complete the calculation.
```

Comment:
accepts an equation and when the equation is ->  accepts an expression and when the expression is